### ISSUE TITLE
node scripts/sync.js index update

### ISSUE THREAD

**Issue by bylucy** [iquidus/explorer#280]
Hello,

node scripts/sync.js index update
update complete (block: 13999)

now blocks 400000
blocks no more sync
how to fix the problem?
please guys help me
tanks

**Comment by bylucy** [comment]
how to fix the problem?

**Comment by wh1tebeard** [comment]
Do you have these cron jobs running?
`*/1 * * * * cd /path/to/explorer && /usr/bin/nodejs scripts/sync.js index update > /dev/null 2>&1
*/2 * * * * cd /path/to/explorer && /usr/bin/nodejs scripts/sync.js market > /dev/null 2>&1
*/5 * * * * cd /path/to/explorer && /usr/bin/nodejs scripts/peers.js > /dev/null 2>&1`

**Comment by uaktags** [comment]
Also, try deleting tmp/index.pid if it exists and let sync.js index update run naturally (or run it yourself to capture some outputs)

**Comment by Zafford** [comment]
Hi, 
my sync.js always ends (with index update) at the same block (3326, no error). If have much more blocks. It seems that stats.count does not get updated with the current last block. Who should update stats,count?

**Comment by uaktags** [comment]
Stats.count gets updated at the beginning of the ``sync.js index update`` functionality. It grabs the count from the daemon. You can add a console.log to output what this count is if you want.

**Comment by Zafford** [comment]
I put in a console.log and it always shows the 3226 independent of the correct number. So it seems that the daemon is not correctly delivering the count. Can I check that?

**Comment by uaktags** [comment]
Well tracing it down:

https://github.com/iquidus/explorer/blob/bf99bb56eb4f0437c59d54a424882c2f839ca521/scripts/sync.js#L134-L138
This is where an initial check on the database goes into play. This merely refers to https://github.com/iquidus/explorer/blob/bf99bb56eb4f0437c59d54a424882c2f839ca521/lib/database.js#L253 which is just a mongo query of ```Stats.findOne``` used to validate that there's a coinstats entry for the coin.

https://github.com/iquidus/explorer/blob/bf99bb56eb4f0437c59d54a424882c2f839ca521/scripts/sync.js#L139 is where we're going to make our calls for getting the total blockcount. Specifically https://github.com/iquidus/explorer/blob/bf99bb56eb4f0437c59d54a424882c2f839ca521/lib/database.js#L605 and Line 606 is where I may put a console.log to see what is coming straight from the daemon.
https://github.com/iquidus/explorer/blob/bf99bb56eb4f0437c59d54a424882c2f839ca521/lib/database.js#L612-L614 is where that entry from the daemon gets saved to Mongo.

https://github.com/iquidus/explorer/blob/bf99bb56eb4f0437c59d54a424882c2f839ca521/scripts/sync.js#L140 is where we end up grabbing that newly saved value.

So what I would do is console.log that 606 line in database.js ``console.log("BlockCount from Coind", count);``, console.log line 141 in sync ``console.log("BlockCount from DB", stats.count);`` and then you can even do a new tab in your browser to https://127.0.0.1:<port>/api/getblockcount. You should now have 3 numbers, and in theory, at the very least, ``BlockCount from Coind`` and the blockcount shown in the new tab should both be the same, while ``BlockCount from DB`` is different.

If that's the case, then we know there's an issue with saving here:
https://github.com/iquidus/explorer/blob/bf99bb56eb4f0437c59d54a424882c2f839ca521/lib/database.js#L612-L614

If there is an issue with saving, I'd start console.log 'ing the outputs.

```
  update_db: function(coin, cb) {
    lib.get_blockcount( function (count) {
      if (!count){
        console.log('Unable to connect to explorer API');
        return cb(false);
      }
      console.log("BlockCount from CoinD", count);
      lib.get_supply( function (supply){
        console.log("Supply from CoinD", supply);
        lib.get_connectioncount(function (connections) {
          console.log("Connections from CoinD", connections);
          Stats.update({coin: coin}, {
            coin: coin,
            count : count,
            supply: supply,
            connections: connections,
          }, function(error) { //note this error parameter is not part of iquidus, added for this example
            if(error){
                console.log(error);
                return cb(false);
            }else{
            console.log("We saved to database")
            return cb(true);
            }
          });
        });
      });
    });
  },
```
This should give you some useful information.

**Comment by Zafford** [comment]
First let me thank you for your support, much appreciated. 
I did all mentioned above and what you wrote "at the very least, BlockCount from Coind and the blockcount shown in the new tab should both be the same, while BlockCount from DB is different." turned out to be correct. Coind and new tab had the same (newest) number and blockcount from db is different (3326). So after logging the output, the problem is with lib.get_supply, it brings the following
"Supply from CoinD undefined" and then "[CastError: Cast to number failed for value "undefined" at path "supply"]"
After commenting out the supply update in stats.,update in database.js, it updated fine.
Then I realized that supply "GETINFO" does not work for my coin, after changing it to "BALANCES" the supply is no longer undefined.
Again, thanks a lot
